File: memory_trainer.js
// memory_trainer.js
#!/usr/bin/env node
const readline = require('readline');
const { program } = require('commander');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function clearScreen() {
    process.stdout.write('\x1b[2J\x1b[0;0H');
}

function generateSequence(length) {
    return Array.from({ length }, () => Math.floor(Math.random() * 10));
}

function displaySequence(seq, showTime) {
    console.log('Remember these numbers:');
    console.log(seq.join(' '));
    setTimeout(() => clearScreen(), showTime * 1000);
    return new Promise(resolve => setTimeout(resolve, showTime * 1000));
}

function getUserInput(length) {
    return new Promise((resolve) => {
        rl.question(`Enter the ${length} numbers (space‑separated): `, (answer) => {
            const parts = answer.trim().split(/\s+/);
            if (parts.length !== length) {
                console.log(`Please enter exactly ${length} numbers.`);
                resolve(getUserInput(length));
                return;
            }
            const nums = parts.map(Number);
            if (nums.some(isNaN)) {
                console.log('Please enter valid integers.');
                resolve(getUserInput(length));
                return;
            }
            resolve(nums);
        });
    });
}

async function playRound(length, showTime) {
    const seq = generateSequence(length);
    await displaySequence(seq, showTime);
    const userSeq = await getUserInput(length);
    return seq.every((v, i) => v === userSeq[i]);
}

async function main() {
    program
        .option('-l, --length <n>', 'Number of digits per sequence', parseInt)
        .option('-d, --difficulty <type>', 'easy, medium, hard')
        .option('-r, --rounds <n>', 'Number of rounds', parseInt, 1)
        .parse(process.argv);

    const opts = program.opts();
    let seqLen;
    if (opts.length) {
        seqLen = opts.length;
    } else if (opts.difficulty) {
        switch (opts.difficulty) {
            case 'easy': seqLen = 4; break;
            case 'medium': seqLen = 6; break;
            case 'hard': seqLen = 8; break;
            default: console.error('Invalid difficulty.'); process.exit(1);
        }
    } else {
        seqLen = 6;
    }
    if (seqLen < 1 || opts.rounds < 1) {
        console.error('Length and rounds must be at least 1.');
        process.exit(1);
    }

    console.log(`\n🧠 Memory Trainer`);
    console.log(`Difficulty: ${opts.difficulty || 'custom'} (${seqLen} digits)`);
    console.log(`Rounds: ${opts.rounds}`);

    let correct = 0;
    const showTime = 2;
    for (let r = 1; r <= opts.rounds; r++) {
        console.log(`\nRound ${r}/${opts.rounds}`);
        const ok = await playRound(seqLen, showTime);
        if (ok) {
            console.log('✅ Correct!');
            correct++;
        } else {
            console.log('❌ Wrong.');
        }
    }
    console.log(`\nScore: ${correct}/${opts.rounds}`);
    rl.close();
}

main().catch(console.error);

File: memory_trainer.cpp
// memory_trainer.cpp
#include <iostream>
#include <vector>
#include <string>
#include <sstream>
#include <cstdlib>
#include <ctime>
#include <thread>
#include <chrono>
#include <getopt.h>

#ifdef _WIN32
#include <windows.h>
#else
#include <unistd.h>
#endif

using namespace std;

void clearScreen() {
#ifdef _WIN32
    system("cls");
#else
    system("clear");
#endif
}

vector<int> generateSequence(int length) {
    vector<int> seq(length);
    for (int i=0; i<length; i++) seq[i] = rand() % 10;
    return seq;
}

void displaySequence(const vector<int>& seq, int showTime) {
    cout << "Remember these numbers:" << endl;
    for (int n : seq) cout << n << " ";
    cout << endl;
    this_thread::sleep_for(chrono::seconds(showTime));
    clearScreen();
}

vector<int> getUserInput(int length) {
    while (true) {
        cout << "Enter the " << length << " numbers (space‑separated): ";
        string line;
        getline(cin, line);
        stringstream ss(line);
        vector<int> nums;
        int num;
        while (ss >> num) nums.push_back(num);
        if ((int)nums.size() != length) {
            cout << "Please enter exactly " << length << " numbers." << endl;
            continue;
        }
        return nums;
    }
}

bool playRound(int length, int showTime) {
    vector<int> seq = generateSequence(length);
    displaySequence(seq, showTime);
    vector<int> userSeq = getUserInput(length);
    return seq == userSeq;
}

int main(int argc, char* argv[]) {
    static struct option long_options[] = {
        {"length", required_argument, 0, 'l'},
        {"difficulty", required_argument, 0, 'd'},
        {"rounds", required_argument, 0, 'r'},
        {0,0,0,0}
    };
    int opt;
    int length = 0;
    string difficulty;
    int rounds = 1;
    while ((opt = getopt_long(argc, argv, "l:d:r:", long_options, nullptr)) != -1) {
        switch (opt) {
            case 'l': length = stoi(optarg); break;
            case 'd': difficulty = optarg; break;
            case 'r': rounds = stoi(optarg); break;
            default:
                cerr << "Usage: memory_trainer --length <n> --difficulty <easy|medium|hard> --rounds <n>" << endl;
                return 1;
        }
    }

    if (length == 0) {
        if (difficulty == "easy") length = 4;
        else if (difficulty == "medium") length = 6;
        else if (difficulty == "hard") length = 8;
        else length = 6;
    }
    if (length < 1 || rounds < 1) {
        cerr << "Length and rounds must be at least 1." << endl;
        return 1;
    }

    srand(time(nullptr));
    cout << "\n🧠 Memory Trainer" << endl;
    cout << "Difficulty: " << (difficulty.empty() ? "custom" : difficulty) << " (" << length << " digits)" << endl;
    cout << "Rounds: " << rounds << endl;

    int correct = 0;
    int showTime = 2;
    for (int r=1; r<=rounds; r++) {
        cout << "\nRound " << r << "/" << rounds << endl;
        if (playRound(length, showTime)) {
            cout << "✅ Correct!" << endl;
            correct++;
        } else {
            cout << "❌ Wrong." << endl;
        }
    }
    cout << "\nScore: " << correct << "/" << rounds << endl;
    return 0;
}
